# ✅ Solved Questions — Class 2
## Variables, Data Types, Comments, Operators, Keywords & Concatenation

This notebook contains **all 80 practice questions from Class 2**, solved in order, section by section.

Every coding question follows this pattern:

```
Question N
   ↓
Code
   ↓
Line-by-line explanation
```

A few conceptual questions (e.g. "explain what `is` checks") get a clear written answer instead of code.

# Section 1 — Practice: Variables

### Question 1 — Create a variable called `city` and store a city name in it.

In [1]:
city = "Karachi"
print(city)

Karachi


**Line-by-line explanation**
1. `city = "Karachi"` creates a variable named `city` and stores the String `"Karachi"` in it.
2. `print(city)` displays the value stored in `city`.

### Question 2 — Create an `age` variable containing an Integer.

In [3]:
age = 18
print(age)

18


**Line-by-line explanation**
1. `age = 25` creates a variable named `age` and stores the whole number `25` — an Integer.
2. `print(age)` displays `25`.

### Question 3 — Create `favorite_number` and print it.

In [4]:
favorite_number = 7
print(favorite_number)

7


**Line-by-line explanation**
1. `favorite_number = 7` creates the variable and stores a number in it (any number works; `7` is used here).
2. `print(favorite_number)` displays `7`.

### Question 4 — Create two variables and print both.

In [6]:
name = "Saad"
age = 18

print(name)
print(age)

Saad
18


**Line-by-line explanation**
1. `name = "Saad"` creates the first variable, storing a String.
2. `age = 25` creates the second variable, storing an Integer.
3. `print(name)` displays `Saad`.
4. `print(age)` displays `25` on its own line.

### Question 5 — Change a variable's value and print the new value.

In [7]:
score = 10
print(score)

score = 20
print(score)

10
20


**Line-by-line explanation**
1. `score = 10` creates `score` with the value `10`.
2. `print(score)` displays `10`.
3. `score = 20` **overwrites** the same variable with a new value — this is legal in Python because variables are not fixed once created.
4. `print(score)` now displays `20`, confirming the value changed.

# Section 2 — Practice: Types of Variables

### Question 1 — Create one String variable.

In [8]:
name = "Saad"
print(name)

Saad


**Line-by-line explanation**
1. `name = "Saad"` stores text inside quotes, which makes it a String.
2. `print(name)` displays `Saad`.

### Question 2 — Create one Integer variable.

In [10]:
age = 18
print(age)

18


**Line-by-line explanation**
1. `age = 25` stores a whole number with no decimal point, which makes it an Integer.
2. `print(age)` displays `25`.

### Question 3 — Create one Float variable.

In [11]:
weight = 60.5
print(weight)

60.5


**Line-by-line explanation**
1. `weight = 60.5` stores a number **with** a decimal point, which makes it a Float.
2. `print(weight)` displays `60.5`.

### Question 4 — Create one Boolean variable.

In [12]:
is_student = True
print(is_student)

True


**Line-by-line explanation**
1. `is_student = True` stores one of Python's two Boolean values, `True`.
2. `print(is_student)` displays `True`.

### Question 5 — Use `type()` to check all four.

In [14]:
name = "Saad"
age = 18
weight = 60.5
is_student = True

print(type(name))
print(type(age))
print(type(weight))
print(type(is_student))

<class 'str'>
<class 'int'>
<class 'float'>
<class 'bool'>


**Line-by-line explanation**
1. The four variables from the previous questions are recreated here together.
2. `type(name)` returns `<class 'str'>`.
3. `type(age)` returns `<class 'int'>`.
4. `type(weight)` returns `<class 'float'>`.
5. `type(is_student)` returns `<class 'bool'>` — confirming all four data types.

# Section 3 — Practice: Strings

### Question 1 — Create a String containing your first name.

In [15]:
first_name = "Saad"
print(first_name)

Saad


**Line-by-line explanation**
1. `first_name = "Saad"` creates a String variable with the first name as its value.
2. `print(first_name)` displays it.

### Question 2 — Create a String using single quotes.

In [16]:
text1 = 'Hello there'
print(text1)

Hello there


**Line-by-line explanation**
1. `'Hello there'` uses single quotes (`'`) instead of double quotes to mark the start and end of the String — Python treats both the same way.
2. `print(text1)` displays `Hello there`.

### Question 3 — Create a String using double quotes.

In [17]:
text2 = "Hello there"
print(text2)

Hello there


**Line-by-line explanation**
1. `"Hello there"` uses double quotes (`"`) — functionally identical to single quotes for a simple String like this.
2. `print(text2)` displays `Hello there`.

### Question 4 — Create a multi-line String using triple quotes.

In [18]:
message = """This is line one.
This is line two.
This is line three."""

print(message)

This is line one.
This is line two.
This is line three.


**Line-by-line explanation**
1. `"""..."""` (triple double quotes) lets a String span multiple lines exactly as typed, including the line breaks inside it.
2. `message` stores all three lines as one single String value.
3. `print(message)` displays the String exactly as written, across three lines.

### Question 5 — Write a String containing quotation marks inside it.

In [19]:
quote = "She said, \"Python is fun!\""
print(quote)

quote2 = 'He said, "Let\'s learn Python."'
print(quote2)

She said, "Python is fun!"
He said, "Let's learn Python."


**Line-by-line explanation**
1. `quote` is written with double quotes as the outer boundary; the double quotes that appear *inside* the text are "escaped" with a backslash (`\\"`) so Python knows they are part of the text, not the end of the String.
2. `print(quote)` displays: `She said, "Python is fun!"`
3. `quote2` shows an alternative approach: using **single** quotes as the outer boundary means the double quotes inside don't need escaping at all — only the apostrophe (`'`) inside needs a backslash (`\\'`).
4. This demonstrates a useful trick: choosing single vs. double quotes as your outer boundary can avoid needing to escape most inner quotation marks.

# Section 4 — Practice: Integer

### Question 1 — Create an Integer containing your age.

In [21]:
age = 18
print(age)

18


**Line-by-line explanation**
1. `age = 25` stores a whole number.
2. `print(age)` displays it.

### Question 2 — Create an Integer with value `100`.

In [22]:
number = 100
print(number)

100


**Line-by-line explanation**
1. `number = 100` creates an Integer variable holding exactly `100`.
2. `print(number)` displays `100`.

### Question 3 — Create a negative Integer.

In [23]:
temperature = -5
print(temperature)

-5


**Line-by-line explanation**
1. A minus sign directly before the number (`-5`) makes it negative — Integers can be positive, negative, or zero.
2. `print(temperature)` displays `-5`.

### Question 4 — Print an Integer and its type.

In [24]:
count = 42
print(count)
print(type(count))

42
<class 'int'>


**Line-by-line explanation**
1. `count = 42` creates the Integer.
2. `print(count)` displays `42`.
3. `print(type(count))` displays `<class 'int'>`, confirming the data type.

### Question 5 — Create two Integer variables and print them.

In [25]:
apples = 5
oranges = 8

print(apples)
print(oranges)

5
8


**Line-by-line explanation**
1. `apples = 5` and `oranges = 8` create two separate Integer variables.
2. The two `print()` calls display `5` and `8` on separate lines.

# Section 5 — Practice: Float

### Question 1 — Create a Float with a decimal value.

In [26]:
height = 5.9
print(height)

5.9


**Line-by-line explanation**
1. `height = 5.9` stores a number with a decimal point, which makes it a Float.
2. `print(height)` displays `5.9`.

### Question 2 — Create a Float with value `10.5`.

In [27]:
value = 10.5
print(value)

10.5


**Line-by-line explanation**
1. `value = 10.5` creates a Float holding exactly `10.5`.
2. `print(value)` displays `10.5`.

### Question 3 — Create a negative Float.

In [28]:
balance = -25.75
print(balance)

-25.75


**Line-by-line explanation**
1. A minus sign before a decimal number (`-25.75`) makes a negative Float — useful for things like a negative account balance.
2. `print(balance)` displays `-25.75`.

### Question 4 — Print a Float and its type.

In [29]:
pi_approx = 3.14
print(pi_approx)
print(type(pi_approx))

3.14
<class 'float'>


**Line-by-line explanation**
1. `pi_approx = 3.14` creates the Float.
2. `print(pi_approx)` displays `3.14`.
3. `print(type(pi_approx))` displays `<class 'float'>`.

### Question 5 — Create two Float variables and print them.

In [30]:
length = 12.5
width = 4.25

print(length)
print(width)

12.5
4.25


**Line-by-line explanation**
1. `length = 12.5` and `width = 4.25` create two Float variables.
2. Both `print()` calls display their respective values on separate lines.

# Section 6 — Practice: Boolean

### Question 1 — Create a Boolean variable called `is_student`.

In [31]:
is_student = True

**Line-by-line explanation**
1. `is_student = True` creates a variable named `is_student`. This line alone doesn't print anything — it just creates the variable, which is exactly what this question asks for.

### Question 2 — Give it the value `True`.

In [32]:
is_student = True
print(is_student)

True


**Line-by-line explanation**
1. `is_student = True` assigns the Boolean value `True` to the variable.
2. `print(is_student)` displays `True`.

### Question 3 — Create another Boolean with `False`.

In [33]:
has_pet = False
print(has_pet)

False


**Line-by-line explanation**
1. `has_pet = False` creates a second Boolean variable with the value `False`.
2. `print(has_pet)` displays `False`.

### Question 4 — Print both Boolean variables.

In [34]:
is_student = True
has_pet = False

print(is_student)
print(has_pet)

True
False


**Line-by-line explanation**
1. Both variables from the earlier questions are recreated here together.
2. The two `print()` calls display `True` and `False` on separate lines.

### Question 5 — Use `type()` to check a Boolean.

In [35]:
is_student = True
print(type(is_student))

<class 'bool'>


**Line-by-line explanation**
1. `is_student = True` creates the Boolean.
2. `print(type(is_student))` displays `<class 'bool'>`, confirming it is a Boolean.

# Section 7 — Practice: Comments

### Question 1 — Write a comment explaining what a variable is.

In [36]:
# A variable is a name that stores a value so it can be reused later.

**Line-by-line explanation**
1. The line starts with `#`, which marks everything after it on that line as a **comment**. Python completely ignores comments when running the program — they exist purely to explain the code to humans.

### Question 2 — Create a variable and add a comment on the same line.

In [37]:
age = 18 # student's age in years
print(age)

18


**Line-by-line explanation**
1. `age = 25` creates the variable as usual.
2. `# student's age in years` is a comment placed **after** the code on the same line — Python runs the `age = 25` part and then ignores everything from `#` onward.
3. `print(age)` displays `25`; the comment has no effect on the program's behavior.

### Question 3 — Write two separate comments.

In [38]:
# This program stores a person's name.
# It then displays the name on the screen.

name = "Saad"
print(name)

Saad


**Line-by-line explanation**
1. Two separate `#` lines are written, each explaining part of what the program does — comments can be stacked on consecutive lines.
2. `name = "Saad"` and `print(name)` are the actual code; the comments above are skipped entirely when the code runs.

### Question 4 — Add comments to a small 3-line program.

In [40]:
# Step 1: store the name
name = "Saad"

# Step 2: store the age
age = 18

# Step 3: display both values
print(name, age)

Saad 18


**Line-by-line explanation**
1. Each comment (`# Step 1...`, `# Step 2...`, `# Step 3...`) documents the purpose of the code line right below it.
2. `print(name, age)` uses a comma (Class 2's concatenation method) to display both values separated by a space: `Saad 25`.

### Question 5 — Add comments explaining important lines of a code example.

In [41]:
# Store the price of one item
price = 250

# Store how many items were bought
quantity = 3

# Calculate the total cost by multiplying price and quantity
total = price * quantity

# Show the final total to the user
print("Total cost:", total)

Total cost: 750


**Line-by-line explanation**
1. Each comment explains *why* the line below it exists, not just *what* it does — good comments describe intent, since the code itself already shows the mechanics.
2. `total = price * quantity` calculates `250 * 3 = 750`.
3. `print("Total cost:", total)` displays `Total cost: 750`, using a comma to combine the label and the number.

# Section 8 — Practice: Arithmetic

### Question 1 — Calculate `20 + 5`.

In [42]:
print(20 + 5)

25


**Line-by-line explanation**
1. `20 + 5` is evaluated first, giving `25`.
2. `print(...)` displays `25`.

### Question 2 — Calculate `20 - 5`.

In [43]:
print(20 - 5)

15


**Line-by-line explanation**
1. `20 - 5` is evaluated first, giving `15`.
2. `print(...)` displays `15`.

### Question 3 — Calculate `20 * 5`.

In [44]:
print(20 * 5)

100


**Line-by-line explanation**
1. `20 * 5` is evaluated first, giving `100`.
2. `print(...)` displays `100`.

### Question 4 — Find the remainder when `20` is divided by `6`.

In [45]:
print(20 % 6)

2


**Line-by-line explanation**
1. `%` is the remainder (modulo) operator: `20 ÷ 6 = 3` remainder `2`.
2. `print(...)` displays `2`.

### Question 5 — Find `2 ** 5`.

In [46]:
print(2 ** 5)

32


**Line-by-line explanation**
1. `**` is the exponent operator: `2 ** 5` means `2` raised to the power `5`, i.e. `2×2×2×2×2 = 32`.
2. `print(...)` displays `32`.

# Section 9 — Practice: Assignment

### Question 1 — Start with `x = 10` and use `+=` to add `5`.

In [47]:
x = 10
x += 5
print(x)

15


**Line-by-line explanation**
1. `x = 10` sets the starting value.
2. `x += 5` means `x = x + 5`, so `x` becomes `15`.
3. `print(x)` displays `15`.

### Question 2 — Start with `x = 10` and use `-=` to subtract `3`.

In [48]:
x = 10
x -= 3
print(x)

7


**Line-by-line explanation**
1. `x = 10` sets the starting value.
2. `x -= 3` means `x = x - 3`, so `x` becomes `7`.
3. `print(x)` displays `7`.

### Question 3 — Start with `x = 4` and use `*=` to multiply by `2`.

In [49]:
x = 4
x *= 2
print(x)

8


**Line-by-line explanation**
1. `x = 4` sets the starting value.
2. `x *= 2` means `x = x * 2`, so `x` becomes `8`.
3. `print(x)` displays `8`.

### Question 4 — Start with `x = 10` and use `/=` to divide by `2`.

In [50]:
x = 10
x /= 2
print(x)

5.0


**Line-by-line explanation**
1. `x = 10` sets the starting value.
2. `x /= 2` means `x = x / 2`, so `x` becomes `5.0` — note it is a Float, since `/` always returns a Float.
3. `print(x)` displays `5.0`.

### Question 5 — Create a variable and use `%=` or `//=` on it.

In [51]:
x = 17
x %= 5
print(x)

y = 17
y //= 5
print(y)

2
3


**Line-by-line explanation**
1. `x %= 5` means `x = x % 5`, the remainder of `17 ÷ 5`, so `x` becomes `2`.
2. `y //= 5` means `y = y // 5`, floor division of `17 ÷ 5` (rounded down), so `y` becomes `3`.
3. Both results are printed on separate lines: `2` then `3`.

# Section 10 — Practice: Comparison

### Question 1 — Compare `10` and `5` using `==`.

In [52]:
print(10 == 5)

False


**Line-by-line explanation**
1. `==` checks equality. `10` is not equal to `5`, so the result is `False`.
2. `print(...)` displays `False`.

### Question 2 — Compare `10` and `5` using `!=`.

In [53]:
print(10 != 5)

True


**Line-by-line explanation**
1. `!=` checks "not equal". `10` is indeed not equal to `5`, so the result is `True`.
2. `print(...)` displays `True`.

### Question 3 — Check whether `10 > 5`.

In [54]:
print(10 > 5)

True


**Line-by-line explanation**
1. `10 > 5` checks if `10` is strictly greater than `5`, which is `True`.
2. `print(...)` displays `True`.

### Question 4 — Check whether `10 <= 5`.

In [55]:
print(10 <= 5)

False


**Line-by-line explanation**
1. `10 <= 5` checks if `10` is less than or equal to `5`. Since `10` is greater, this is `False`.
2. `print(...)` displays `False`.

### Question 5 — Create two numbers and use all six comparison operators.

In [56]:
a = 10
b = 7

print(a == b)
print(a != b)
print(a > b)
print(a < b)
print(a >= b)
print(a <= b)

False
True
True
False
True
False


**Line-by-line explanation**
1. `a = 10` and `b = 7` set up two numbers to compare.
2. `a == b` → `False` (not equal).
3. `a != b` → `True` (they are different).
4. `a > b` → `True` (10 is greater).
5. `a < b` → `False` (10 is not less than 7).
6. `a >= b` → `True`.
7. `a <= b` → `False`.

# Section 11 — Practice: Logical

### Question 1 — Evaluate `True and True`.

In [57]:
print(True and True)

True


**Line-by-line explanation**
1. `and` needs **both** sides to be `True`. Since both are `True`, the result is `True`.

### Question 2 — Evaluate `True and False`.

In [58]:
print(True and False)

False


**Line-by-line explanation**
1. `and` needs both sides `True`. Since the second is `False`, the whole result is `False`.

### Question 3 — Evaluate `False or True`.

In [59]:
print(False or True)

True


**Line-by-line explanation**
1. `or` only needs **one** side to be `True`. Since the second value is `True`, the result is `True`.

### Question 4 — Use `not` with `True`.

In [60]:
print(not True)

False


**Line-by-line explanation**
1. `not` reverses a Boolean value. `not True` becomes `False`.

### Question 5 — Create two Boolean variables and use `and`, `or`, and `not`.

In [61]:
is_sunny = True
is_weekend = False

print(is_sunny and is_weekend)
print(is_sunny or is_weekend)
print(not is_sunny)

False
True
False


**Line-by-line explanation**
1. `is_sunny = True` and `is_weekend = False` set up two Booleans.
2. `is_sunny and is_weekend` → `False` (needs both `True`).
3. `is_sunny or is_weekend` → `True` (only needs one `True`).
4. `not is_sunny` → `False` (reverses `True`).

# Section 12 — Practice: Identity

### Question 1 — Create a list called `a`.

In [62]:
a = [1, 2, 3]
print(a)

[1, 2, 3]


**Line-by-line explanation**
1. `a = [1, 2, 3]` creates a list (Class 4) with three numbers.
2. `print(a)` displays `[1, 2, 3]`.

### Question 2 — Assign `b = a` and test `a is b`.

In [63]:
a = [1, 2, 3]
b = a

print(a is b)

True


**Line-by-line explanation**
1. `b = a` does **not** create a new list — `b` now points to the exact same list object as `a` in memory.
2. `a is b` checks whether both names refer to the same object. Since they do, the result is `True`.

### Question 3 — Create another list with the same values and compare it using `is`.

In [64]:
a = [1, 2, 3]
c = [1, 2, 3]

print(a == c)
print(a is c)

True
False


**Line-by-line explanation**
1. `c = [1, 2, 3]` creates a **brand-new, separate** list that happens to contain the same values as `a`.
2. `a == c` → `True`, because `==` compares the *values* inside the lists, and they match.
3. `a is c` → `False`, because `is` compares *identity* (are they literally the same object in memory?) — and they are not, since `c` was built independently.

### Question 4 — Test `is not` with two variables.

In [65]:
a = [1, 2, 3]
c = [1, 2, 3]

print(a is not c)

True


**Line-by-line explanation**
1. `a is not c` checks the opposite of `is`. Since `a` and `c` are different objects (as shown in Question 3), the result is `True`.

### Question 5 — Explain in your own words what `is` checks.

**Answer:**

`is` checks whether two variables point to the **exact same object in memory** — not just whether their values look the same. Two separate lists can contain identical values (so `==` says `True`) while still being two different objects in memory (so `is` says `False`).

# Section 13 — Practice: Membership

### Question 1 — Check whether `2` is in `[1, 2, 3]`.

In [66]:
print(2 in [1, 2, 3])

True


**Line-by-line explanation**
1. `in` checks whether a value exists inside a list. `2` is present, so the result is `True`.

### Question 2 — Check whether `5` is in `[1, 2, 3]`.

In [67]:
print(5 in [1, 2, 3])

False


**Line-by-line explanation**
1. `5` does not appear anywhere in `[1, 2, 3]`, so the result is `False`.

### Question 3 — Check whether 'a' is in 'Saad'.

In [68]:
print('a' in 'Saad')

True


**Line-by-line explanation**
1. `in` also works on Strings — it checks whether the smaller string is a substring of the larger one. The letter `'a'` does appear in `'Saad'`, so the result is `True`.

### Question 4 — Check whether 'z' is not in 'Saad'.

In [69]:
print('z' not in 'Saad')

True


**Line-by-line explanation**
1. `'z'` does **not** appear in `'Saad'`, so `'z' not in 'Saad'` evaluates to `True`.

### Question 5 — Create your own list and test `in` and `not in`.

In [70]:
fruits = ["apple", "banana", "mango"]

print("banana" in fruits)
print("grape" in fruits)
print("grape" not in fruits)

True
False
True


**Line-by-line explanation**
1. `fruits = [...]` creates a list of three fruit names.
2. `"banana" in fruits` → `True`, since `"banana"` is in the list.
3. `"grape" in fruits` → `False`, since `"grape"` is missing.
4. `"grape" not in fruits` → `True`, the opposite of the previous check.

# Section 14 — Practice: Bitwise

### Question 1 — Calculate `5 & 3`.

In [71]:
print(5 & 3)

1


**Line-by-line explanation**
1. `5` is `101` in binary, `3` is `011`. `&` (AND) keeps a `1` only where **both** bits are `1`: `001` → `1`.

### Question 2 — Calculate `5 | 3`.

In [72]:
print(5 | 3)

7


**Line-by-line explanation**
1. `|` (OR) keeps a `1` where **at least one** bit is `1`: `101 | 011 = 111` → `7`.

### Question 3 — Calculate `5 ^ 3`.

In [73]:
print(5 ^ 3)

6


**Line-by-line explanation**
1. `^` (XOR) keeps a `1` where the bits are **different**: `101 ^ 011 = 110` → `6`.

### Question 4 — Calculate `5 << 1`.

In [74]:
print(5 << 1)

10


**Line-by-line explanation**
1. `<<` shifts bits left by 1 position, which doubles the value: `5 << 1 = 10`.

### Question 5 — Calculate `5 >> 1`.

In [75]:
print(5 >> 1)

2


**Line-by-line explanation**
1. `>>` shifts bits right by 1 position, which halves the value (rounded down): `5 >> 1 = 2`.

# Section 15 — Practice: Keywords

### Question 1 — Write five Python keywords you remember.

**Answer:**

Five examples: `if`, `else`, `while`, `for`, `return`. (Other correct answers include `def`, `class`, `True`, `False`, `None`, `import`, `try`, `except`.)

### Question 2 — Is `if` a keyword?

In [76]:
import keyword
print(keyword.iskeyword("if"))

True


**Line-by-line explanation**
1. `import keyword` brings in Python's built-in `keyword` module, which knows every reserved word.
2. `keyword.iskeyword("if")` checks whether `"if"` is a reserved keyword, and returns `True` — confirming `if` is indeed a keyword.

### Question 3 — Why can't keywords normally be used as ordinary variable names?

**Answer:**

Keywords are **reserved** by Python because they have a special, fixed meaning in the language's own grammar (for example, `if` always starts a conditional statement). If Python allowed `if = 5`, it would become impossible for Python to tell whether `if` means "the start of a condition" or "a variable I just created" — so keywords are blocked from being used as ordinary names to avoid this ambiguity.

### Question 4 — Identify the keywords in a statement containing `if` and `else`.

In [85]:
age = 20

if age >= 18:
    print("Adult")
else:
    print("Minor")

Adult


**Line-by-line explanation**
1. In this statement, `if` and `else` are the two **keywords** — reserved words that control which branch of code runs.
2. `age`, `print` are not keywords (`age` is a normal variable name; `print` is a built-in function name, which is different from a keyword — it *can* technically be reassigned, though doing so is bad practice).
3. The condition `age >= 18` is `True` (since `age` is `20`), so `"Adult"` is printed.

### Question 5 — Write three keywords from this class without looking at the notes.

**Answer:**

Three examples: `class`, `try`, `return`. (Any three genuine Python keywords are correct — the point of this question is active recall, not a specific list.)

# Section 16 — Practice: Concatenation

### Question 1 — Join `"Hello"` and `"World"` using `+`.

In [86]:
print("Hello" + "World")
print("Hello" + " " + "World")

HelloWorld
Hello World


**Line-by-line explanation**
1. `"Hello" + "World"` joins the two Strings directly, giving `HelloWorld` with no space between them.
2. Adding a separate `" "` String in between (`"Hello" + " " + "World"`) gives the more natural-looking `Hello World`.

### Question 2 — Print a name and age using commas.

In [87]:
name = "Saad"
age = 18

print("Name:", name, "Age:", age)

Name: Saad Age: 18


**Line-by-line explanation**
1. Commas inside `print()` let Python display multiple values (even different types — a String and an Integer here) separated by spaces automatically, with no need for `str()`.
2. This displays: `Name: Saad Age: 25`.

### Question 3 — Print a name and age using `+` and `str()`.

In [88]:
name = "Saad"
age = 18

print("Name: " + name + " Age: " + str(age))

Name: Saad Age: 18


**Line-by-line explanation**
1. `+` requires every piece being joined to already be a String.
2. `name` is already a String, so it joins directly.
3. `age` is an Integer, so it must be wrapped in `str(age)` first to convert it to text before `+` can join it — without `str()`, this would raise a `TypeError`.
4. The result is the same sentence: `Name: Saad Age: 25`.

### Question 4 — Print a name and age using an f-string.

In [89]:
name = "Saad"
age = 18

print(f"Name: {name} Age: {age}")

Name: Saad Age: 18


**Line-by-line explanation**
1. The `f` before the opening quote marks this as an f-string.
2. `{name}` and `{age}` are automatically replaced with the current values of those variables — no `str()` conversion needed, since f-strings handle any type automatically.
3. This displays the same sentence: `Name: Saad Age: 25`.

### Question 5 — Create a greeting using `.format()`.

In [90]:
name = "Saad"
greeting = "Hello, {}! Welcome to Python.".format(name)

print(greeting)

Hello, Saad! Welcome to Python.


**Line-by-line explanation**
1. `{}` inside the String is a placeholder.
2. `.format(name)` fills that placeholder with the value of `name`, in order.
3. `print(greeting)` displays: `Hello, Saad! Welcome to Python.`

---

## 🎉 Class 2 — All 80 Questions Solved

Every practice question from Class 2 — Variables, Data Types, Strings, Integers, Floats, Booleans, Comments, all 7 Operator categories, Keywords, and Concatenation — has now been solved with code and line-by-line explanations.



---

## 👨‍💻 About This Course

This **free Python course** was designed and developed by **Muhammad Saad** as a contribution toward making programming education more accessible. I created it to share what I’ve learned with students and beginners who want to learn Python but may not have access to paid courses.

The goal is simple: **to make learning Python easier, accessible, and beginner-friendly.** If this course helps someone learn a new skill or take their first step into programming, then this initiative has achieved its purpose.

If you found this class helpful, feel free to share it with someone who might benefit from it.